### PR TITLE
Keep test agent state out of commits

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Broad test-generation runs currently create and retain `.testagent/` inside the repository, which allows fully autonomous workflows to commit internal research, planning, and status files alongside generated tests.

This change introduces a shared `<TESTAGENT_DIR>` contract across the test-generation skill and its sub-agents. Intermediate state files now prefer host-provided scratch storage, fall back to worktree-specific Git metadata, and use OS temporary storage. The invariant is that these files are non-stageable: they must not be version-controlled workspace content or appear in `git status`. The eval suite detects stageable `research.md`, `plan.md`, or `status.md` files regardless of parent directory.

A conclusive Claude Sonnet 4.6 evaluation exposed a separate completeness over-correction in broad suites. The pytest judge found that the baseline had "materially broader reported behavioral and edge-case coverage (56 tests versus 43)," and the xUnit judge preferred its "meaningful decimal-pricing and reservation coverage beyond the minimum." The requirement matrix had become an implicit ceiling. Broad workflows now treat named requirements as the floor, then add one mutation-relevant case for every distinct equivalence partition or invariant, while still rejecting raw test-count padding.

## Related issue

N/A

## Validation

- `dotnet run --project eng\skill-validator\src\SkillValidator.csproj -p:CopilotSkipCliDownload=true --no-restore -- check --plugin .\plugins\dotnet-test` - passed
- `python eng\eval-quality\check_eval_quality.py` - passed with existing repository warnings
- `git diff --check` - passed
- Intermediate-state grader probe - broad runs require external state; focused runs reject both Git-metadata and stageable state files, including ignored files, while `node_modules` is excluded by pathspec
- Workspace-integrity setup probe - one eval-root Git repository tracks the fixture deletions, retains the synthetic target, and leaves no nested repository
- Run `33734244414` - merge-head evaluation passed with Claude Sonnet 4.6 at 7W/1T/1L (p=0.035) and GPT-5.6 Luna at 9W/0T/0L (p=0.002); fresh evaluation requested for `4b73d614b`

## Checklist

- [x] I searched existing issues and pull requests to avoid duplicates.
- [x] I kept this pull request focused and avoided unrelated refactors.
- [x] I added or updated tests, evals, or documentation when changing skill or agent behavior.
- [x] I updated CODEOWNERS when adding or moving owned content. N/A: no owned content was added or moved.
- [x] I updated all marketplace manifests when plugin metadata changed. N/A: plugin metadata was unchanged.
- [x] I updated `eng/known-domains.txt` for any new external domains referenced by skill content. N/A: no external domains were added.

### CODE REVIEW

File: plugins/dotnet-test/skills/code-testing-agent/SKILL.md [hunk 1]
```diff
@@ -138,15 +142,28 @@ Generate unit tests for [path or description of what to test], following the [un
...
+same Research → Plan → Implement sequence inline, resolve `TESTAGENT_DIR` as
+described below, create the internal artifacts there, and apply the same
+completion contract.
+
+For broad scope, resolve one absolute `TESTAGENT_DIR` before creating state:
+
+1. Prefer a host-provided session artifact or scratch directory.
+2. Otherwise, in a Git worktree use the absolute form of
+   `git rev-parse --git-path testagent`; this is worktree-specific Git metadata
+   and cannot be staged.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply plugins/dotnet-test/skills/code-testing-agent/SKILL.md]: Agreed. Step 2 now explicitly uses `--path-format=absolute`, removing ambiguity around the absolute `TESTAGENT_DIR` contract.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/agents/code-testing-generator.agent.md [hunk 2]
```diff
@@ -35,15 +35,30 @@ You coordinate test generation using the Research-Plan-Implement (RPI) pipeline.
...
+Before writing code, read the language-specific base extension. Reuse it for the whole run; sub-agents must not independently reload the same reference unless they need a section that was not captured in the research document.
+
+For Single pass and Iterative strategies, resolve one absolute `TESTAGENT_DIR`
+before invoking any sub-agent:
+
+1. Prefer a host-provided session artifact or scratch directory when one is
+   available.
+2. Otherwise, in a Git worktree use the absolute form of
+   `git rev-parse --git-path testagent`. This stores state in worktree-specific
+   Git metadata, which cannot be staged or committed.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-generator.agent.md]: Good catch. Updated the generator instructions with `git rev-parse --path-format=absolute --git-path testagent`, which explicitly returns an absolute path.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/skills/code-testing-extensions/extensions/dotnet.md [hunk 3]
```diff
@@ -101,7 +101,7 @@ not part of the test assembly and must never be reported as generated coverage.
 
 A new `.csproj` is **invisible** to `dotnet test <solution>`, to `dotnet test` run from the repo root, and to any CI/benchmark harness until it is added to the solution. Run `dotnet sln add` *immediately* after creating the project as part of Step 3 ("Register Test Project with Build System") — do not defer it to a later step.
 
-1. Use the exact solution or solution-filter target identified in `.testagent/research.md` or `.testagent/plan.md` — do not search for or substitute a different `.sln`, `.slnx`, or `.slnf` target.
+1. Use the exact solution or solution-filter target identified in the research or plan document under `TESTAGENT_DIR` — do not search for or substitute a different `.sln`, `.slnx`, or `.slnf` target.
 2. If that target is a `.sln` or `.slnx`, run `dotnet sln <solution> add <test-project.csproj>`.
 3. If the target is a `.slnf` (solution filter), also ensure the new project is included in the filter; adding only to the underlying `.sln` may not be enough for test discovery.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply plugins/dotnet-test/skills/code-testing-extensions/extensions/dotnet.md:106]: Agreed. The placeholder now uses `<TESTAGENT_DIR>`, and the builder, tester, and linter guidance now requires the caller to provide an exact command, excerpt, or absolute document path.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by AbhitejJohn — APPROVED** [review]
Thanks for tightening this up. The core `TESTAGENT_DIR` approach works, and I found two issues worth following up on:

1. **Proven:** The eval check passes with `research.md` at the repository root (`exit 0`) and fails only for a directory named `.testagent` (`exit 1`). It also no longer verifies that broad runs create the required pipeline artifacts.
2. **Proven:** The linter's only caller does not pass the researched lint command or `TESTAGENT_DIR`, unlike the builder and tester callers.

The eval-quality checks and plugin validation pass. Neither issue blocks the core commit-safety goal, so I am approving this PR.

_(Copilot, commenting on Abhitej's behalf.)_

File: tests/dotnet-test/code-testing-agent/eval.yaml [hunk 4]
```diff
@@ -298,19 +295,18 @@ stimuli:
...
         config:
           pattern: \|\s*Requirement\s*\|\s*Evidence\s*\|
-      - type: file-exists
-        config:
-          path: .testagent/research.md
-      - type: file-exists
+      - type: run-command
         config:
-          path: .testagent/plan.md
+          command: sh -c "state_dir=\"$(git rev-parse --path-format=absolute --git-path testagent)\" && test -f \"$state_dir/research.md\" && test -f \"$state_dir/plan.md\" && test -f \"$state_dir/status.md\" && test -z \"$(find . -type d -name .git -prune -o -name .testagent -print -quit)\""
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/code-testing-agent/eval.yaml]: Agreed. All eval guards now use constant-time checks for the workspace-root `.testagent` path, including broken symlinks, so they never traverse `node_modules` or other dependency trees.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by Evangelink — COMMENTED** [review]
/evaluate

File: tests/dotnet-test/code-testing-agent/eval.yaml [hunk 5]
```diff
@@ -49,20 +49,19 @@ stimuli:
...
         config:
           pattern: \|\s*Requirement\s*\|\s*Evidence\s*\|
-      - type: file-exists
-        config:
-          path: .testagent/research.md
-      - type: file-exists
+      - type: run-command
         config:
-          path: .testagent/plan.md
+          command: state_dir="$(git rev-parse --path-format=absolute --git-path testagent)" && test -f "$state_dir/research.md" && test -f "$state_dir/plan.md" && test -f "$state_dir/status.md" && test ! -e .testagent && test ! -L .testagent
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/code-testing-agent/eval.yaml]: Accepted. Rather than weakening artifact verification for a non-Git workspace, all seven broad eval environments now initialize their root as Git, matching the autonomous workflow and making the worktree-metadata path deterministic.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by azat-msft — APPROVED** [review]

File: plugins/dotnet-test/agents/code-testing-generator.agent.md [hunk 6]
```diff
@@ -87,25 +102,25 @@ Delegate to the `code-testing-researcher` subagent with this task:
 ```text
 runSubagent({
   agent: "code-testing-researcher",
-  prompt: "Research [REQUESTED SCOPE] at [PATH] for test generation. Produce a bounded target inventory, existing test conventions, source-to-test pairs, dependencies only for those targets, and exact build/test/discovery commands. Do not inventory unrelated source files."
+  prompt: "Research [REQUESTED SCOPE] at [PATH] for test generation. Write the research document to [ABSOLUTE TESTAGENT_DIR]/research.md. Produce a bounded target inventory, existing test conventions, source-to-test pairs, dependencies only for those targets, and exact build/test/discovery commands. Do not inventory unrelated source files."
```
Comment (JanKrivanek) [inline plugins/dotnet-test/agents/code-testing-generator.agent.md]: We seem to mix 3 different refences to test agent temp dir:
- `TESTAGENT_DIR`
-  `<TESTAGENT_DIR>`
-  `[ABSOLUTE TESTAGENT_DIR]`
  
Should we just stick to single one?

File: plugins/dotnet-test/agents/code-testing-generator.agent.md [hunk 7]
```diff
@@ -152,19 +167,28 @@ Additional self-review heuristics (still required, even when running the skills)
...
+6. Stop only when every feasible checklist item and distinct behavioral
+   partition is covered and the stated target is met. Do not recursively expand
+   into unrelated files or add equivalent cases merely to raise test count.
+7. If this step added or modified tests, re-run the full Step 7 pre-completion gate (`test-gap-analysis` + `assertion-quality` + prompt-scenario coverage) on those tests before reporting completion.
+
+For Single pass and Iterative strategies, write `<TESTAGENT_DIR>/status.md` after
 the final review and validation. Record the completed checklist, commands and
 results, quality findings, fixes, and any explicit blockers. Direct strategy
-keeps this evidence in the final response and must not create `.testagent/`.
+keeps this evidence in the final response and must not create pipeline state.
```
Comment (JanKrivanek) [inline plugins/dotnet-test/agents/code-testing-generator.agent.md]: "pipeline state" sounds it can be confusing. Should we call it more expressively? E.g. "intermediate state files"?

File: plugins/dotnet-test/agents/code-testing-generator.agent.md [hunk 8]
```diff
@@ -224,9 +249,9 @@ All state is stored in `.testagent/` folder:
 6. **Scoped builds during phases, full build at the end** — build specific test projects during implementation for speed; run a full-workspace non-incremental build after all phases to catch cross-project errors
 7. **No environment-dependent tests** — mock all external dependencies; never call external URLs, bind ports, or depend on timing
 8. **Fix assertions, don't skip tests** — when tests fail, read production code and fix the expected value; never `[Ignore]` or `[Skip]`
-9. **Retain `.testagent/` through completion** — keep the research, plan, and final status available as auditable pipeline evidence. Do not delete them automatically; if the repository should not commit agent state, advise the user to add `.testagent/` to `.gitignore` after reporting the result.
+9. **Keep agent state out of commits** — retain research, plan, and final status in `TESTAGENT_DIR` through completion, but never create `.testagent/` in the workspace, stage agent state, or modify `.gitignore` for it. Before reporting, inspect the working-tree changes and confirm they contain only requested deliverables and required manifest edits.
```
Comment (JanKrivanek) [inline plugins/dotnet-test/agents/code-testing-generator.agent.md]: I believe we should not refer to '.testagent/' here. We shaould again mention that `<TESTAGENT_DIR>` should not be created in workspace.

In the new version of .md - this is the first mention of '.testagent/' - so it can get confusing, why would it even be created? While agent can decide to create e.g. "AgentDir" (which still follows the previous instructions fine) and then if it would add it to workspace - it would not break this clausule here.

File: plugins/dotnet-test/agents/code-testing-implementer.agent.md [hunk 9]
```diff
@@ -62,14 +64,21 @@ For each test file in your phase:
...
 
 - **Existing test files are append-only.** When growing an existing test file, insert new test methods/cases at the end of the relevant class/describe-block/module. Do not reformat, reorder, rename, or remove any existing line — even whitespace-only churn counts as a destructive edit.
-- **Do not modify non-test source files.** If a class, method, or symbol is hard to test (sealed, internal, no seam, tightly coupled), record the gap in `.testagent/plan.md` as a follow-up. Do not edit production code to make it testable as part of test generation — that is the scope of the `testability-migration` agent, not this one.
+- **Do not modify non-test source files.** If a class, method, or symbol is hard to test (sealed, internal, no seam, tightly coupled), record the gap in `<TESTAGENT_DIR>/plan.md` as a follow-up. Do not edit production code to make it testable as part of test generation — that is the scope of the `testability-migration` agent, not this one.
+- **Keep pipeline state non-stageable.** Use only the absolute `TESTAGENT_DIR`
+  supplied by the caller for research, plan, or status updates. Never create
+  `.testagent/` in the workspace or stage pipeline state.
```
Comment (JanKrivanek) [inline plugins/dotnet-test/agents/code-testing-implementer.agent.md]: Again - .testagent seems superflous here (similar as we'd say "never create 'foo-bar'" - never mentioned before and then suddenly just mentioned as that it should not be created)

File: tests/dotnet-test/code-testing-agent/eval.yaml [hunk 10]
```diff
@@ -468,7 +474,7 @@ stimuli:
           timeout: 1m
       - type: run-command
         config:
-          command: sh -c "test -z \"$(find . -type d -name .testagent -print -quit)\""
+          command: sh -c "test ! -e .testagent && test ! -L .testagent"
```
Comment (JanKrivanek) [inline tests/dotnet-test/code-testing-agent/eval.yaml]: Since the state dir can now be named differently (and arbitrarily) - we should not test for the folder, but for actual .md files - research.md, plan.md, status.md

File: plugins/dotnet-test/agents/code-testing-generator.agent.md [hunk 11]
```diff
@@ -152,19 +167,28 @@ Additional self-review heuristics (still required, even when running the skills)
...
+   Sweep each bounded target API for still-unproved observable equivalence
+   partitions and invariants: identity/empty/singleton/interior inputs, exact
+   and immediately adjacent boundaries, invalid partitions, and ordering,
+   monotonicity, rollover, capacity, truncation, or state properties implied by
+   the implementation. Add one mutation-relevant case per distinct partition;
+   consolidate sibling inputs in parameterized or table-driven tests.
+6. Stop only when every feasible checklist item and distinct behavioral
+   partition is covered and the stated target is met. Do not recursively expand
+   into unrelated files or add equivalent cases merely to raise test count.
+7. If this step added or modified tests, re-run the full Step 7 pre-completion gate (`test-gap-analysis` + `assertion-quality` + prompt-scenario coverage) on those tests before reporting completion.
```
Comment (JanKrivanek) [inline plugins/dotnet-test/agents/code-testing-generator.agent.md:188]: Technically this is very unrelated improvement and would be better in separate PR. but that is a very minor nit

(applies to couple other similar changes accross this PR)

**Review by JanKrivanek — APPROVED** [review]
Overall it's a good change and I'm good with it. I just have couple comments I want to be strongly considered ;-)

File: plugins/dotnet-test/agents/code-testing-generator.agent.md (hunk 11 above)
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-generator.agent.md:188]: I considered splitting this, but am keeping it here because the user explicitly requested the Claude completeness follow-up on this PR, and it changes the same test-generation skill and eval surface. It has also been independently validated by multiple passing evaluations.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/agents/code-testing-implementer.agent.md (hunk 9 above)
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-implementer.agent.md]: Accepted. The implementer now enforces the generic `<TESTAGENT_DIR>` stageability contract without referring to the legacy `.testagent` name.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: tests/dotnet-test/code-testing-agent/eval.yaml (hunk 10 above)
Comment (Evangelink) [reply tests/dotnet-test/code-testing-agent/eval.yaml]: Accepted. The eval now detects stageable `research.md`, `plan.md`, or `status.md` files through Git regardless of their parent directory, while separately verifying the required external files.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/agents/code-testing-generator.agent.md (hunk 7 above)
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-generator.agent.md]: Accepted. The guidance now calls these "intermediate state files," which is more explicit about what research, plan, and status documents are.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/agents/code-testing-generator.agent.md (hunk 6 above)
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-generator.agent.md]: Accepted. All conceptual references now consistently use `<TESTAGENT_DIR>`, including the sub-agent prompt examples.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: plugins/dotnet-test/agents/code-testing-generator.agent.md (hunk 8 above)
Comment (Evangelink) [reply plugins/dotnet-test/agents/code-testing-generator.agent.md]: Accepted. The rule is now directory-name-independent: `<TESTAGENT_DIR>` and its intermediate state files must never be placed in version-controlled workspace content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by Evangelink — COMMENTED** [review]
/evaluate

File: tests/dotnet-test/code-testing-agent/eval.yaml [hunk 12]
```diff
@@ -230,6 +232,7 @@ stimuli:
           dest: fixtures/python-workspace-integrity
       commands:
         - cd fixtures/python-workspace-integrity && git init -q && git config maintenance.auto false && git config gc.auto 0 && git -c user.email=eval@example.com -c user.name=eval add -A && git -c user.email=eval@example.com -c user.name=eval commit -qm baseline && rm -rf core io tests_real
+        - git init -q
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/code-testing-agent/eval.yaml]: Accepted. The workspace-integrity stimulus now initializes only the eval-root repository, commits the fixture from that root, and then removes the tracked subtrees, so the stageable-file guard sees the entire workspace.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by Evangelink — COMMENTED** [review]
/evaluate

**Review by AbhitejJohn — APPROVED** [review]
Thanks for the thorough update. The new recursive `git ls-files` check addresses the nested intermediate-state case I found earlier.

I still see two conditional evaluator gaps: the broad graders require state under `git rev-parse --git-path testagent` although the skill prefers host scratch when available, and `--exclude-standard` can hide intermediate files covered by `.gitignore`. The workspace-integrity case also resolves state from a different Git root when the agent works inside the nested fixture repository.

These are worth following up, but I am approving this revision.

_(Copilot, commenting on Abhitej's behalf.)_

**Comment by AbhitejJohn** [comment]
Correction: the PR head advanced to `d9bdbe1cd2` during review submission. That commit now uses one Git root for the workspace-integrity eval, so that concern is resolved. The host-scratch location mismatch and `--exclude-standard` blind spot remain as non-blocking follow-up items.

_(Copilot, commenting on Abhitej's behalf.)_

File: tests/dotnet-test/code-testing-agent/eval.yaml [hunk 13]
```diff
@@ -428,7 +433,9 @@ stimuli:
             --include='*.cs'
             && ! grep -Rq 'OrderPricing' fixtures/sdk-xunit-orders/tests
             --include='*.cs'
-            && test -z \"$(find . -type d -name .testagent -print -quit)\""
+            && test -z \"\$(git ls-files --cached --others --exclude-standard
+            -- ':(glob)**/research.md' ':(glob)**/plan.md'
+            ':(glob)**/status.md')\""
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/code-testing-agent/eval.yaml]: Accepted. I removed the state check from nested `sh -c` entirely and made it a direct grader, so substitution is unambiguous. Focused graders now also reject Git-metadata state, include ignored intermediate files, and exclude only `node_modules` via pathspec.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by AbhitejJohn — APPROVED** [review]
